// muse-ui/lib/List/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListItemAfterText = exports.ListItemSubTitle = exports.ListItemTitle = exports.ListItemContent = exports.ListAction = exports.ListItem = exports.List = undefined;

require('../styles/components/list.less');

var _theme = require('../theme');

var _theme2 = _interopRequireDefault(_theme);

var _theme3 = require('./theme');

var _theme4 = _interopRequireDefault(_theme3);

var _utils = require('../utils');

var _List = require('./List');

var _List2 = _interopRequireDefault(_List);

var _ListItem = require('./ListItem');

var _ListItem2 = _interopRequireDefault(_ListItem);

var _ListAction = require('./ListAction');

var _ListAction2 = _interopRequireDefault(_ListAction);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.List = _List2.default;
exports.ListItem = _ListItem2.default;
exports.ListAction = _ListAction2.default;
var ListItemContent = exports.ListItemContent = (0, _utils.createSimpleFunctional)('mu-item-content', 'div', 'mu-list-item-content');
var ListItemTitle = exports.ListItemTitle = (0, _utils.createSimpleFunctional)('mu-item-title', 'div', 'mu-list-item-title');
var ListItemSubTitle = exports.ListItemSubTitle = (0, _utils.createSimpleFunctional)('mu-item-sub-title', 'div', 'mu-list-item-sub-title');
var ListItemAfterText = exports.ListItemAfterText = (0, _utils.createSimpleFunctional)('mu-item-after-text', 'span', 'mu-list-item-after-text');

_List2.default.install = function (Vue) {
  Vue.component(_List2.default.name, _List2.default);
  Vue.component(_ListItem2.default.name, _ListItem2.default);
  Vue.component(_ListAction2.default.name, _ListAction2.default);
  Vue.component(ListItemContent.name, ListItemContent);
  Vue.component(ListItemTitle.name, ListItemTitle);
  Vue.component(ListItemSubTitle.name, ListItemSubTitle);
  Vue.component(ListItemAfterText.name, ListItemAfterText);
};

_theme2.default.addCreateTheme(_theme4.default);
exports.default = _List2.default;
